[DPE-2983] HA network cut

juditnovak opened the pull request:
NOTE to Reviewers:

Plese take a look at "Usage Instructions" in [this comment](https://github.com/canonical/opensearch-dashboards-operator/pull/22#issuecomment-2092721246) before reviewing this PR

Shotrly:
When reviewing changs, in the Files (left-hand side) tab
 - close the `libs` tab (nothig to review there)
 - close the `tests/integration/opensearch-operator` tab (nothing to review here)
 
 Furthemore, you may find review by commit handy.

juditnovak commented on src/events/tls.py:
```diff
@@ -116,6 +118,11 @@ def _on_certificate_expiring(self, _: EventBase) -> None:
+    def _on_config_changed(self, event: EventBase):
+        """If system configuration (such as IP) changes, certs have to be re-issued."""
+        if self.charm.state.unit_server.tls and not self.charm.tls_manager.certificate_valid():
+            self._on_certificate_expiring(event)
```
Otherwise: 
```
curl: (60) SSL: no alternative certificate subject name matches target host name '10.77.44.85'
```

Mehdi-Bendriss commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_network_cut_application(ops_test: OpsTest, request):
+    """SIGKILLs leader process and checks recovery + re-election."""
```
wrong docstring?

juditnovak replied:
Fixed

Mehdi-Bendriss commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+            ha_helpers.restore_unit_network(machine_name)
+        except CalledProcessError:  # in case it was already cleaned up
+            pass
+
```
similar let's check each unit's IP changed

juditnovak replied:
Done.

Mehdi-Bendriss commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+
+    await asyncio.sleep(RESTART_DELAY * 2)
+    await extra_secure_wait_for_idle(ops_test)
+
```
lets check no IP changed

juditnovak replied:
Done.

Mehdi-Bendriss replied:
check missing

juditnovak replied:
As said in https://github.com/canonical/opensearch-dashboards-operator/pull/22#discussion_r1591129618

This is independent from the functionality of the charm/application.

This is LXD + Juju working together, isn't it...?
Anyway, adding the check

Mehdi-Bendriss replied:
If the IP address change does not happen or happens when not expected to, our tests are effectively not testing what we wanted them to test

juditnovak replied:
Added as requestd.

Mehdi-Bendriss reviewed (commented):
Thanks Judit, I have a few not too big remarks (references to ZK should be dropped)

juditnovak commented on src/events/tls.py:
```diff
@@ -116,6 +118,11 @@ def _on_certificate_expiring(self, _: EventBase) -> None:
 
+    def _on_config_changed(self, event: EventBase):
+        """If system configuration (such as IP) changes, certs have to be re-issued."""
+        if self.charm.state.unit_server.tls and not self.charm.tls_manager.certificate_valid():
```
Isn't that a sufficient indicator that the IP in the currently stored certificate is different from the current IP of the node...?

juditnovak replied:
..at the moment when receiving a `config-changed` event?

(Which only triggers on 3 distinct cases: Actual configuration change, right after startup, AND on IP change...?
https://ops.readthedocs.io/en/latest/#ops.ConfigChangedEvent)

I'm asking because it's a non-trivial change, and except if strong beenfid, I'd prefer to skip it.

HA tests have been pretty elaborate.

juditnovak replied:
Moreover. Config check shouldn't go int the TLS event handlers. 

From TLS perspective we have one duty to do on a `config-changed` eevnt. We have to check if the certificates are still relevant -- and if they are not, we must request new ones. That is the scope of the TLS `config-changed` event handler.

Now note that on `config-changed` events the main `reconcile()` method is being also triggered. In case the config is to be changed (or may have changed) that's the execution to be triggered.

Overall: your concern regarding the config is valid, however it's not the right place (and time) yet to worry about that one here :-)

Mehdi-Bendriss replied:
This is tracked here. https://github.com/canonical/opensearch-dashboards-operator/issues/27

juditnovak replied:
Issue closed as I'm implementing it already at the moment.

juditnovak replied:
Responded.

juditnovak commented on src/events/tls.py:
```diff
@@ -116,6 +118,11 @@ def _on_certificate_expiring(self, _: EventBase) -> None:
+    def _on_config_changed(self, event: EventBase):
+        """If system configuration (such as IP) changes, certs have to be re-issued."""
+        if self.charm.state.unit_server.tls and not self.charm.tls_manager.certificate_valid():
+            self._on_certificate_expiring(event)
```
TLS lib acts the same on 'invalid' and 'expired' certs. 

Also: it's kinda on purpose not to distinguish further amont ghe events. It's a policy.

juditnovak replied:
As discussed live: should be sufficient the way it is -- and I hope I can carry on with the same poliicy in the future, however we'll see what time brings :-)

juditnovak replied:
Responded.

juditnovak commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+@pytest.mark.skip_if_deployed
+@pytest.mark.abort_on_fail
+async def test_deploy_active(ops_test: OpsTest):
+    """Tests that the charm deploys safely, without DNS resolution from LXD dnsmasq."""
```
Copy-paste

juditnovak replied:
Fixed, we don't need this.

juditnovak commented on tests/integration/ha/helpers.py:
```diff
@@ -0,0 +1,458 @@
+    subprocess.check_call(enable_dnsmasq.split())
+
+
+def cut_unit_network(machine_name: str) -> None:
```
`cut` is IP cahnge, `throttle` is no IP change.

I can add it in the docstring (I thought it was there though but I may have missed)

juditnovak replied:
I have taken a careful attention to call the actual test functions according to your requested naming (i.e. indicating in the test function name `ip_change`, `no_ip_change`.

Regarding helper functions, I would like to stick to the current naming -- while I made sure to very clearly indicate in the docstrings what is the impact on the IP address.

I may not be the only one to who can distinguish a lot easier between `cut` and `throttle` words than seeking for a small `no` tag within long function names. This is the rationale why I would I find the current naming handy, and I would like to stick to it. While assuring clarity (docstrings) and following your naming on the actual "interface" i.e. pytest executable test functions. 

I hope this could be a good compromise for both of us :-)

phvalguima commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+
+
+async def extra_secure_wait_for_idle(ops_test: OpsTest):
+    """Particularly strong, specific wait_for_idle time period.
```
We should move to the `wait_until`, as the opensearch itself is doing.

juditnovak replied:
This one seemed reliable for this particular case. https://github.com/juju/python-libjuju/blob/d4f667766655e1eadbb5fd8e6cb40eb9e9451039/juju/model.py#L2829-L2832

As I understand  `wait_for_at_least_units` isn't only checking for the number of those units, but also for the `idle` state of them. Pls correct if I'm wrong.

TBH this change I may prefer as a separate PR. The simple reason is that HA tests are _tedious_ and take a long time to debug.

Now we have a stable solution here. I'd perefer to have a "less fancy" stable thing, than perhaps a broken fancy one.

How about me addressint this in a separate PR, and we sort out all the rest on this one?

phvalguima replied:
I noticed you replaced parts with `block_until`. I think this is fine.

phvalguima commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_network_throttle_leader(ops_test: OpsTest, request):
+    """SIGKILLs leader process and checks recovery + re-election."""
```
Wrong docstring?

juditnovak replied:
Fixed.

phvalguima commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_network_cut_application(ops_test: OpsTest, request):
+    """SIGKILLs leader process and checks recovery + re-election."""
```
Wrong docstring?

juditnovak replied:
Fixed.

phvalguima commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_network_cut_leader(ops_test: OpsTest, request):
+    """SIGKILLs leader process and checks recovery + re-election."""
```
Wrong docstring

juditnovak replied:
Fixed.

phvalguima commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+@pytest.mark.skip_if_deployed
+@pytest.mark.abort_on_fail
+async def test_deploy_active(ops_test: OpsTest):
+    """Tests that the charm deploys safely, without DNS resolution from LXD dnsmasq."""
```
Wrong docstring?

juditnovak replied:
Fixed.

phvalguima commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+#
+#
+# @pytest.fixture()
+# async def no_lxd_dnsmasq():
```
Guess this check is not being done?

juditnovak replied:
Fixed.

phvalguima commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_network_throttle_application(ops_test: OpsTest, request):
+    """SIGKILLs leader process and checks recovery + re-election."""
```
Wrong docstring

juditnovak replied:
Fixed

phvalguima commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_network_cut_leader_https(ops_test: OpsTest, request):
+    """SIGKILLs leader process and checks recovery + re-election."""
```
Wrong docstring

juditnovak replied:
Fixed

phvalguima commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_network_throttle_leader_https(ops_test: OpsTest, request):
+    """SIGKILLs leader process and checks recovery + re-election."""
```
Wrong docstring

juditnovak replied:
Fixed

phvalguima commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_network_cut_application_https(ops_test: OpsTest, request):
+    """SIGKILLs leader process and checks recovery + re-election."""
```
Wrong docstring

juditnovak replied:
Fixed

phvalguima reviewed (changes requested):
Pretty big PR, I focused on changes in: `tests/integration/ha/*`

Overall, I suggest we move to [OpenSearch's wait_until](https://github.com/canonical/opensearch-operator/blob/46bc4a8f46905228a576c97d29de1572dc141bdd/tests/integration/ha/test_ha_networking.py#L95). The advantage of `wait_until` is printing the entire juju status.

When we have a network cut, we should expect the targeted unit to move to `unknown/lost` status which I think `wait_idle` from libjuju won't show.

juditnovak commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_network_cut_leader(ops_test: OpsTest, request):
```
I distinguish between `cut` (IP change) and `throttle` (no IP change).

Additionally I can add a specific docstring. IS that enough or you insist on the function name?

juditnovak replied:
Closing it as a duplicate, see https://github.com/canonical/opensearch-dashboards-operator/pull/22#discussion_r1590413831

juditnovak commented on tests/integration/ha/helpers.py:
```diff
@@ -0,0 +1,458 @@
+    stop=stop_after_attempt(60),
+    reraise=True,
+)
+def srvr(host: str) -> dict:
```
Yes, I thought I may need the code but I can remove it now :-)

juditnovak replied:
In fact I can't -- I'll need (and modify) it in an upcoming task (namely HA process failure).

BTW all this can be removed once _really_ not needed in https://warthogs.atlassian.net/browse/DPE-3954

juditnovak commented:
@phvalguima Evaluating your comment 

> Pretty big PR

```
kaboca@pegasus:~/repos/duplicated/opensearch-dashboards-operator$ git log --oneline -10
6762835 (HEAD -> DPE-2983_ha_network_cut, origin/DPE-2983_ha_network_cut) TLS operator v136
9cf625d HA tests
8a5d2fe Removing unnecessary unittest
e213a06 Certificate refresh on IP change (+ small adjustments)
946e85e Generic tests cleanup from local Opensearch charm copy
519e6cf Removing local copy of Opensearch
2450340 Updating libs
ef6cbbc (origin/main, origin/HEAD, main) Merge pull request #19 from canonical/4033_tls_workflow_cleanup
6ea9bb7 (origin/4033_tls_workflow_cleanup) fixup! Changes on PR request
6a228ab Changes on PR request
kaboca@pegasus:~/repos/duplicated/opensearch-dashboards-operator$ git log ef6cbbc..519e6cf --numstat --pretty="%H" | awk 'NF==3 {added+=$1; deleted+=$2} NF==1 {commit++} END {printf("total lines added: +%d\ntotal lines deleted: -%d\ntotal commits: %d\n", added, deleted, commit)}'
total lines added: +154
total lines deleted: -26017
total commits: 2
```
While I agree that the PR "seems" big, on a closer look what you find is:
 - 25 000+ lines were _removed_ -- **no reveiw needed**
   - local copy of the Opensearch charm is not needed in test space anymore, since https://github.com/canonical/opensearch-operator/pull/216 got merged
   - easy to ignore in review space by closing the `opensearch-operator` file pull-down list, since no review needed on this
   - added in isolation as a separate commit
 - outdated charm libs got updated -- **no reveiw needed**
   - added in isolation as a separate commit (two actually, as I consider TLS libs v13 "experimental" :-) )
   - easy to ignore by closing the `libs/charms` file pull-down listt, as o review is needed on this
   - note: the charm has no libs on it's own so far
- the core add-on in this PR is the HA testing, which indeed is a big, very repetititve module (as I guess it's normal for such test modules)

Overall, defending my case: I don't think the PR is really "big" once taking a closer look.
While I DO agree that I should have made all these details clear for reviewers in the **Description**. I'll do that next time.
But the level of "complexity" involved in this PR I think is kinda normal given the weight of pipeline executions, that additional gradening (libs, or removing garbage) is handiest to be added alongside normal tasks...

juditnovak commented on src/charm.py:
```diff
@@ -136,11 +134,15 @@ def reconcile(self, event: EventBase) -> None:
-            clear_status(self.unit, MSG_TLS_CONFIG)
 
+        # Request new certificates if IP changed
+        if self.state.unit_server.tls and self.state.unit_server.tls:
```
Bah, my bad, I couldn't read my own code :facepalm: 

I aplogize and thank you very much for pointing it out. Fixed.

juditnovak commented on tests/integration/ha/helpers.py:
```diff
@@ -0,0 +1,458 @@
+    """
+    cut_network_command = f"lxc config device add {machine_name} eth0 none"
+    subprocess.check_call(cut_network_command.split())
+
```
See custom blocking instead suited to the specific network cut performed.

juditnovak commented on tests/integration/ha/helpers.py:
```diff
@@ -0,0 +1,458 @@
+    subprocess.check_call(limit_set_command.split())
+    limit_set_command = f"lxc config device set {machine_name} eth0 limits.priority=10"
+    subprocess.check_call(limit_set_command.split())
+
```
See custom block in test helper functions

juditnovak commented on tests/integration/ha/helpers.py:
```diff
@@ -0,0 +1,458 @@
+    limit_set_command = f"lxc config device set {machine_name} eth0 limits.priority="
+    subprocess.check_call(limit_set_command.split())
+    restore_unit_network(machine_name=machine_name)
+
```
As above

juditnovak commented on tests/integration/ha/helpers.py:
```diff
@@ -0,0 +1,458 @@
+    """
+    restore_network_command = f"lxc config device remove {machine_name} eth0"
+    subprocess.check_call(restore_network_command.split())
+
```
As above

juditnovak commented:
@phvalguima I don't think blocking using a test helper or not may be necessary :-)

As explained [above](https://github.com/canonical/opensearch-dashboards-operator/pull/22#discussion_r1588994034): I would _not_ like to use the Opensearch `wait_until()` method/module.

I used it before on Opensearch, and I found it particularly difficult to use when debugging a particular issue.

Instead of status output, I have added particularly strict checks on the PR with blocking and parsing status output making sure that expected behavior is carried out -- while aiming to keep it all simple. I payed special attantion to the `unknown/lost` status that you've highlighted earlier.

Given the already existing complexity, the [Juju bug](https://bugs.launchpad.net/juju/+bug/2064876) discovered (in particularly time-consuming ways -- since only appeared on the pipeline) I would like to avoid introducing any additional layer of complexity whenever avoidable.

juditnovak commented on src/core/models.py:
```diff
@@ -52,6 +53,9 @@ def update(self, items: dict[str, str]) -> None:
             self._relation_data.update(update_fields)
+            # Hack to ensure a relation-changed event -- no problem if secret-changed is unreliable
+            # see https://bugs.launchpad.net/juju/+bug/2064876
+            self._relation_data.update({"_relation_data_updated": str(uuid.uuid4())})
```
I quite insist on this one.

Took a VERY long time of my life until the corresponding bug got filed... :-( 

With the "fix" we're safe -- as confirmed by the tests that were failing earlier.

I agree that this is not ideal. However since our goal is to deliver a stable charm -- seems necessary.

juditnovak replied:
Ignore, not the right solution -- working on a better on.

juditnovak replied:
(Also, one shortcoming: it only works when >1 nodes. If it's a single server, then it's the next `update-status` event that will populate the changes. I can add it to the commit above the workaround -- however if that's the only change left on this PR, with the 1.5 new pipeline, then I may rather do it on the next PR like [Scaling tests](https://github.com/canonical/opensearch-dashboards-operator/pull/3) ;-)  Not gonna get forgotten -- how I wish I could fortget this one :-) )

juditnovak commented on src/charm.py:
```diff
@@ -130,24 +132,37 @@ def reconcile(self, event: EventBase) -> None:
+            # NOTE: We never clean up MSG_RECONFIG, as it MUST be replaced by rolling_ops statuses
+            # (Otherwise the server is out-of-date)
+            # It is to ensure continuality of non-active status as needed
+            self.unit.status = MaintenanceStatus(MSG_RECONFIG)
```
I don't insist on this (as VERY soon the charm goes into `rolling_ops` restart phases anyawy.)

If this is raising a debate I rather prefer to get rid of if.

Mehdi-Bendriss replied:
We can get rid of it.

juditnovak replied:
Done.

juditnovak commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+
+    await asyncio.sleep(RESTART_DELAY * 2)
+    await extra_secure_wait_for_idle(ops_test)
+
```
Done.

Mehdi-Bendriss replied:
check missing

juditnovak replied:
TBH at this point I guess we're checking if LXD and Juju work as expected...

But fine, I can add an additional check to verify....

juditnovak replied:
Done.

Mehdi-Bendriss commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,381 @@
+
+    logger.info("Waiting for Juju to detect new IPs...")
+    await ops_test.model.block_until(
+        lambda: not any(ip in ha_helpers.get_hosts_from_status(ops_test).values() for ip in ips),
```
This is risky, I don't think we have the guarantee that a previously allocated IP will not be reused for **another** unit. 
We should rather test the change of IPs for each unit `unit1_old_ip VS unit1_new_ip`

juditnovak replied:
The same check is considered safe on other Data Platform charms (perhaps for a low likelyhood on regularly running pipelines). Neverthless we can make it even more secure.

juditnovak replied:
Changed as requested.

Mehdi-Bendriss commented on tests/integration/helpers.py:
```diff
@@ -144,26 +135,36 @@ def access_dashboard(
 
 
-async def access_all_dashboards(ops_test: OpsTest, relation_id: int, https: bool = False):
+async def access_all_dashboards(
```
```suggestion
async def access_all_dashboards_units(
```

juditnovak replied:
May I keep internal helper functions names as they are -- except when particularly strong reason for otherwise?

Please note that we're modifying an 1.5 h pipeline where the smallest typo may take 45mins to notice.
(And that we were hoping to merge this code last week already...)

Mehdi-Bendriss reviewed (commented):
Thanks Judit. Looks good, there is a couple of checks missing (opened comments) 
* Could you remove references to ZooKeeper from the helpers?

juditnovak commented:
@Mehdi-Bendriss 

> Could you remove references to ZooKeeper from the helpers?

Primarily if any garbage left here is to be cleaned up in here: https://warthogs.atlassian.net/browse/DPE-3954

I would rather suggest to concentrate on the HA network tests making sure that further develoment (cleanup) can think of it as solid. Typically when requesting reveiws -- that casually become possible earlier than 1.5h pipelines verification each time. 
Given that we can assume HA tests being solid, i.e reveiws CAN be requested while HA tests sitll running -- speeding up feedback cycles of 80+ comments (as on this PR).

This PR is partcularly heavy, and speidng 4-5 days on hunting down the related bug https://bugs.launchpad.net/juju/+bug/2064876 (only appearing on heavy pipelines) doubled the already significant weight. I see little value in helper function names or docstring cosmetics added on the top of the above -- especially from a moral perspective.

juditnovak commented on tests/integration/ha/test_network_cut.py:
```diff
@@ -0,0 +1,426 @@
+
+    await asyncio.sleep(RESTART_DELAY * 2)
+    await extra_secure_wait_for_idle(ops_test)
+
```
Done
